### ISSUE TITLE
test-case: general or amigaos4 specific #4

### ISSUE THREAD

**Issue by kas1e** [ptitSeb/gl4es#51]
There is source of morph3d whic shows 3D morphing objects.
I tested that over legacy opengl, all is fine. By default some figure fly around, and by case 1,2,3,4,5 i can change geometry. Also by space and enter can change the things, too. With GL4ES it show black screen.

So, there is source:
http://kas1e.mikendezign.com/aos4/gl4es/tests/testcase_4/morph3d_gl4es.c

There is debug output:
http://kas1e.mikendezign.com/aos4/gl4es/tests/testcase_4/morph3d_debug.txt

From logs i see that shaders fail, will try to check if they compiles over our shader's compiler at all..

**Comment by ptitSeb** [comment]
This is a problem with gl4es, the shader generated by `fpe_shader.c` for double side lighting seems wrong.
I'll fix that.

**Comment by kas1e** [comment]
Checked latest commit, still same:

LIBGL: FPE Vertex shader compile failed: ERROR: Code generation failed.
Error log:
INTERNAL ERROR: Variable is missing its CGVars:
	       152: OpLoad: FloatVector3: tmp152 << back_ss
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: (Vector/Matrix)TimesScalar instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: Dot instruction's source parameters are missing
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp159 >> BackColor
INTERNAL ERROR: Variable is missing its CGVars:
	       245: OpLoad: FloatVector3: tmp245 << back_ss
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: (Vector/Matrix)TimesScalar instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: Dot instruction's source parameters are missing
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	    
glCreateShader(GL_FRAGMENT_SHADER)
glShaderSource(257, 1, 0x6af51f00, 0x0)
Shader source// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000000, texture=00000000, texformat=00000000 point=0
varying vec4 Color;
varying vec4 BackColor;
void main() {
vec4 fColor = (gl_FrontFacing)?Color:BackColor;
gl_FragColor = fColor;
}:
 (FPEShader generated)
New Shader source:
#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000000, texture=00000000, texformat=00000000 point=0
varying vec4 Color;
varying vec4 BackColor;
void main() {
vec4 fColor = (gl_FrontFacing)?Color:BackColor;
gl_FragColor = fColor;
}
glCompileShader(257)
glGetShaderiv(257, 0x8B81, 0x65e4e4b4)
glGetShaderInfoLog(257, 1000, 0x0, 0x65e4e4b8)
LIBGL: FPE Fragment shader compile failed: ERROR: Code generation failed.
Error log:
Loading from built-in variables is unsupported:
	offset: unset, size: 1, Bool gl_FrontFacing BuiltIn(FrontFacing)
INTERNAL ERROR: CompositeConstruct instruction is incomplete:
	        20: OpCompositeConstruct: BoolVector4: tmp20 << tmp13, tmp13, tmp13, tmp13
INTERNAL ERROR: CompositeConstruct instruction is incomplete:
	        20: OpCompositeConstruct: BoolVector4: tmp20 << tmp13, tmp13, tmp13, tmp13
INTERNAL ERROR: CompositeConstruct instruction is incomplete:
	        20: OpCompositeConstruct: BoolVector4: tmp20 << tmp13, tmp13, tmp13, tmp13
INTERNAL ERROR: CompositeConstruct instruction is incomplete:
	        20: OpCompositeConstruct: BoolVector4: tmp20 << tmp13, tmp13, tmp13, tmp13
OpSelect instruction isn't implemented yet.
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp21 >> gl_FragColor

**Comment by ptitSeb** [comment]
Ah, it's another issue. This time, it's an issue in `shaderconv.c` that hasn't translate `gl_FrontFacing`... I'll fix that soon.

**Comment by kas1e** [comment]
And the error in vertex shader still remain the same (i.e. not only frament one fail because of gl_FrontFacing, but vertex one fail the same as before fix. But probably its all related together :)

**Comment by ptitSeb** [comment]
Mmmm, in fact, the OGLES2 drivers should allow the use of `gl_FrontFacing`. This built-in exist in OpenGLES2, I don't see any reason to reject the shader.

**Comment by kas1e** [comment]
Strange, even with latest commits i have 1:1 the same errors as before.

Is that example works for you ?

I will try for now to compile shaders manually via our internal-shader-compiler-checker

**Comment by ptitSeb** [comment]
Yeah, the example works for me.

Vertex shader is
```
#version 100
precision mediump float;
precision mediump int;
uniform highp mat4 _gl4es_ModelViewProjectionMatrix;
uniform highp mat3 _gl4es_NormalMatrix;
attribute highp vec4 _gl4es_Vertex;
attribute highp vec3 _gl4es_Normal;
struct _gl4es_LightModelProducts
{
   vec4 sceneColor;
};
uniform _gl4es_LightModelProducts _gl4es_FrontLightModelProduct;
uniform _gl4es_LightModelProducts _gl4es_BackLightModelProduct;
// FPE_Shader generated
// ** Vertex Shader **
// ligthting=1 (twosided=1, separate=0, color_material=0)
// secondary=0, planes=000000
// texture=00000000, point=0
varying vec4 Color;
struct _gl4es_FPELightSourceParameters1
{
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
   highp float constantAttenuation;
   highp float linearAttenuation;
   highp float quadraticAttenuation;
};
struct _gl4es_FPELightSourceParameters0
{
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
};
struct _gl4es_LightProducts
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
};
uniform highp float _gl4es_FrontMaterial_shininess;
uniform highp float _gl4es_FrontMaterial_alpha;
uniform highp float _gl4es_BackMaterial_shininess;
uniform highp float _gl4es_BackMaterial_alpha;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;
uniform _gl4es_LightProducts _gl4es_BackLightProduct_0;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_1;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_1;
uniform _gl4es_LightProducts _gl4es_BackLightProduct_1;
varying vec4 BackColor;

void main() {
vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);
gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;
// ColorMaterial On/Off=0 Front = 0 Back = 0
Color = _gl4es_FrontLightModelProduct.sceneColor;
BackColor = _gl4es_BackLightModelProduct.sceneColor;
highp float att;
highp float spot;
highp vec3 VP;
highp float lVP;
highp float nVP;
highp vec3 aa,dd,ss;
highp vec3 hi;
highp vec3 back_aa,back_dd,back_ss;
// light 0 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_0.position.xyz);
aa = _gl4es_FrontLightProduct_0.ambient.xyz;
back_aa = _gl4es_BackLightProduct_0.ambient.xyz;
nVP = dot(normal, VP);
dd = nVP * _gl4es_FrontLightProduct_0.diffuse.xyz*step(0.0, nVP);
back_dd = -nVP * _gl4es_BackLightProduct_0.diffuse.xyz * step(0.0, -nVP);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (pow(lVP, _gl4es_FrontMaterial_shininess)*_gl4es_FrontLightProduct_0.specular.xyz)*step(0.0, nVP)*step(0.0, lVP);
ss = (pow(-lVP, _gl4es_BackMaterial_shininess)*_gl4es_BackLightProduct_0.specular.xyz)*step(0.0, -nVP)*step(0.0, -lVP);
Color.rgb += att*(aa+dd+ss);
BackColor.rgb += att*(back_aa+back_dd+back_ss);
// end of light 0
// light 1 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_1.position.xyz);
aa = _gl4es_FrontLightProduct_1.ambient.xyz;
back_aa = _gl4es_BackLightProduct_1.ambient.xyz;
nVP = dot(normal, VP);
dd = nVP * _gl4es_FrontLightProduct_1.diffuse.xyz*step(0.0, nVP);
back_dd = -nVP * _gl4es_BackLightProduct_1.diffuse.xyz * step(0.0, -nVP);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (pow(lVP, _gl4es_FrontMaterial_shininess)*_gl4es_FrontLightProduct_1.specular.xyz)*step(0.0, nVP)*step(0.0, lVP);
ss = (pow(-lVP, _gl4es_BackMaterial_shininess)*_gl4es_BackLightProduct_1.specular.xyz)*step(0.0, -nVP)*step(0.0, -lVP);
Color.rgb += att*(aa+dd+ss);
BackColor.rgb += att*(back_aa+back_dd+back_ss);
// end of light 1
Color.a = _gl4es_FrontMaterial_alpha;
Color.rgb = clamp(Color.rgb, 0., 1.);
BackColor.rgb = clamp(BackColor.rgb, 0., 1.);
BackColor.a = _gl4es_BackMaterial_alpha;
// texturing
}
```
and Fragmen is
```
#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000000, texture=00000000, texformat=00000000 point=0
varying vec4 Color;
varying vec4 BackColor;
void main() {
vec4 fColor = (gl_FrontFacing)?Color:BackColor;
gl_FragColor = fColor;
}
```

**Comment by kas1e** [comment]
Ok, seems our shader's compiler have another limitations.. That what i have when trying to compile fragment shader:

W3DNShaderInfo - Get shader information

Shader: frag.spv
Compiling frag.spv failed (23) with error: shader compilation failed due to errors
Log:
ERROR: Code generation failed.
Error log:
Loading from built-in variables is unsupported:
	offset: unset, size: 1, Bool gl_FrontFacing BuiltIn(FrontFacing)
INTERNAL ERROR: OpBranchConditional instruction's source parameters are missing


 
And that what i have when compile Vertex shader:

W3DNShaderInfo - Get shader information

Shader: vert.spv
Compiling vert.spv failed (23) with error: shader compilation failed due to errors
Log:
ERROR: Code generation failed.
Error log:
INTERNAL ERROR: Variable is missing its CGVars:
	       152: OpLoad: FloatVector3: tmp152 << back_ss
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: (Vector/Matrix)TimesScalar instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: Dot instruction's source parameters are missing
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp159 >> BackColor
INTERNAL ERROR: Variable is missing its CGVars:
	       245: OpLoad: FloatVector3: tmp245 << back_ss
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: (Vector/Matrix)TimesScalar instruction's source parameters are missing
INTERNAL ERROR: I/FAdd instruction's source parameters are missing
INTERNAL ERROR: Dot instruction's source parameters are missing
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp252 >> BackColor 


Seems another 2 bug reports, but at this time to the Warp3D's shader's compiler ?

**Comment by ptitSeb** [comment]
Wait, why I don't `back_ss` and still see `ss` on that shader.

**Comment by ptitSeb** [comment]
New vertex shader:
```
#version 100
precision mediump float;
precision mediump int;
uniform highp mat4 _gl4es_ModelViewProjectionMatrix;
uniform highp mat3 _gl4es_NormalMatrix;
attribute highp vec4 _gl4es_Vertex;
attribute highp vec3 _gl4es_Normal;
struct _gl4es_LightModelProducts
{
   vec4 sceneColor;
};
uniform _gl4es_LightModelProducts _gl4es_FrontLightModelProduct;
uniform _gl4es_LightModelProducts _gl4es_BackLightModelProduct;
// FPE_Shader generated
// ** Vertex Shader **
// ligthting=1 (twosided=1, separate=0, color_material=0)
// secondary=0, planes=000000
// texture=00000000, point=0
varying vec4 Color;
struct _gl4es_FPELightSourceParameters1
{
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
   highp float constantAttenuation;
   highp float linearAttenuation;
   highp float quadraticAttenuation;
};
struct _gl4es_FPELightSourceParameters0
{
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
};
struct _gl4es_LightProducts
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
};
uniform highp float _gl4es_FrontMaterial_shininess;
uniform highp float _gl4es_FrontMaterial_alpha;
uniform highp float _gl4es_BackMaterial_shininess;
uniform highp float _gl4es_BackMaterial_alpha;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;
uniform _gl4es_LightProducts _gl4es_BackLightProduct_0;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_1;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_1;
uniform _gl4es_LightProducts _gl4es_BackLightProduct_1;
varying vec4 BackColor;

void main() {
vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);
gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;
// ColorMaterial On/Off=0 Front = 0 Back = 0
Color = _gl4es_FrontLightModelProduct.sceneColor;
BackColor = _gl4es_BackLightModelProduct.sceneColor;
highp float att;
highp float spot;
highp vec3 VP;
highp float lVP;
highp float nVP;
highp vec3 aa,dd,ss;
highp vec3 hi;
highp vec3 back_aa,back_dd,back_ss;
// light 0 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_0.position.xyz);
aa = _gl4es_FrontLightProduct_0.ambient.xyz;
back_aa = _gl4es_BackLightProduct_0.ambient.xyz;
nVP = dot(normal, VP);
dd = nVP * _gl4es_FrontLightProduct_0.diffuse.xyz*step(0.0, nVP);
back_dd = -nVP * _gl4es_BackLightProduct_0.diffuse.xyz * step(0.0, -nVP);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (pow(lVP, _gl4es_FrontMaterial_shininess)*_gl4es_FrontLightProduct_0.specular.xyz)*step(0.0, nVP)*step(0.0, lVP);
back_ss = (pow(-lVP, _gl4es_BackMaterial_shininess)*_gl4es_BackLightProduct_0.specular.xyz)*step(0.0, -nVP)*step(0.0, -lVP);
Color.rgb += att*(aa+dd+ss);
BackColor.rgb += att*(back_aa+back_dd+back_ss);
// end of light 0
// light 1 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_1.position.xyz);
aa = _gl4es_FrontLightProduct_1.ambient.xyz;
back_aa = _gl4es_BackLightProduct_1.ambient.xyz;
nVP = dot(normal, VP);
dd = nVP * _gl4es_FrontLightProduct_1.diffuse.xyz*step(0.0, nVP);
back_dd = -nVP * _gl4es_BackLightProduct_1.diffuse.xyz * step(0.0, -nVP);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (pow(lVP, _gl4es_FrontMaterial_shininess)*_gl4es_FrontLightProduct_1.specular.xyz)*step(0.0, nVP)*step(0.0, lVP);
back_ss = (pow(-lVP, _gl4es_BackMaterial_shininess)*_gl4es_BackLightProduct_1.specular.xyz)*step(0.0, -nVP)*step(0.0, -lVP);
Color.rgb += att*(aa+dd+ss);
BackColor.rgb += att*(back_aa+back_dd+back_ss);
// end of light 1
Color.a = _gl4es_FrontMaterial_alpha;
Color.rgb = clamp(Color.rgb, 0., 1.);
BackColor.rgb = clamp(BackColor.rgb, 0., 1.);
BackColor.a = _gl4es_BackMaterial_alpha;
// texturing
}
```

**Comment by kas1e** [comment]
That one compiles fine.

Only fragment's one in question then..

**Comment by ptitSeb** [comment]
For the fragment error, it seems it (the Warp3D shader compiler) doesn't like an `if` (well here a ternary) that use a built-in uniform.
Maybe I can do a special workaround to copy that builtin in a regular variable...

**Comment by kas1e** [comment]
In meantime will write to authors of GLES2 and warp3d drivers about. That should be fixed on our side anyway as well even if workoround possible (if you will do so, plz also add it like amigao4, so we can remove it later).

**Comment by ptitSeb** [comment]
Can you try this: in `src/gl/fpe_shader.c` line 735
change
`sprintf(buff, "vec4 fColor = %s;\n", twosided?"(gl_FrontFacing)?Color:BackColor":"Color");`

by
```
if(twosided) { ShadAppend("bool _gl4es_FrontFacing = gl_FrontFacing;"); }
sprintf(buff, "vec4 fColor = %s;\n", twosided?"(_gl4es_FrontFacing)?Color:BackColor":"Color");
```

**Comment by ptitSeb** [comment]
or just try to compile this:
```
#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000000, texture=00000000, texformat=00000000 point=0
varying vec4 Color;
varying vec4 BackColor;
void main() {
bool _gl4es_FrontFacing = gl_FrontFacing;
vec4 fColor = (_gl4es_FrontFacing)?Color:BackColor;
gl_FragColor = fColor;
}
````

**Comment by kas1e** [comment]
Nope, same error about "loading from build int variable is unsupported".. Probably still bugreport is must ?

**Comment by ptitSeb** [comment]
Yes, you need to make a bug report.

I assume that the built-in is not availble in the Fragment Shader, but should be in the Vertex shader.
Are those two shader compiling fine!
Vertex
```
#version 100
precision mediump float;
precision mediump int;
uniform highp mat4 _gl4es_ModelViewProjectionMatrix;
uniform highp mat3 _gl4es_NormalMatrix;
attribute highp vec4 _gl4es_Vertex;
attribute highp vec3 _gl4es_Normal;
struct _gl4es_LightModelProducts
{
   vec4 sceneColor;
};
uniform _gl4es_LightModelProducts _gl4es_FrontLightModelProduct;
uniform _gl4es_LightModelProducts _gl4es_BackLightModelProduct;
// FPE_Shader generated
// ** Vertex Shader **
// ligthting=1 (twosided=1, separate=0, color_material=0)
// secondary=0, planes=000000
// texture=00000000, point=0
varying vec4 Color;
struct _gl4es_FPELightSourceParameters1
{
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
   highp float constantAttenuation;
   highp float linearAttenuation;
   highp float quadraticAttenuation;
};
struct _gl4es_FPELightSourceParameters0
{
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
};
struct _gl4es_LightProducts
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
};
uniform highp float _gl4es_FrontMaterial_shininess;
uniform highp float _gl4es_FrontMaterial_alpha;
uniform highp float _gl4es_BackMaterial_shininess;
uniform highp float _gl4es_BackMaterial_alpha;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;
uniform _gl4es_LightProducts _gl4es_BackLightProduct_0;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_1;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_1;
uniform _gl4es_LightProducts _gl4es_BackLightProduct_1;
varying vec4 BackColor;
varying bool _gl4es_FrontFacing;

void main() {
_gl4es_FrontFacing = gl_FrontFacing;
vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);
gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;
// ColorMaterial On/Off=0 Front = 0 Back = 0
Color = _gl4es_FrontLightModelProduct.sceneColor;
BackColor = _gl4es_BackLightModelProduct.sceneColor;
highp float att;
highp float spot;
highp vec3 VP;
highp float lVP;
highp float nVP;
highp vec3 aa,dd,ss;
highp vec3 hi;
highp vec3 back_aa,back_dd,back_ss;
// light 0 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_0.position.xyz);
aa = _gl4es_FrontLightProduct_0.ambient.xyz;
back_aa = _gl4es_BackLightProduct_0.ambient.xyz;
nVP = dot(normal, VP);
dd = nVP * _gl4es_FrontLightProduct_0.diffuse.xyz*step(0.0, nVP);
back_dd = -nVP * _gl4es_BackLightProduct_0.diffuse.xyz * step(0.0, -nVP);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (pow(lVP, _gl4es_FrontMaterial_shininess)*_gl4es_FrontLightProduct_0.specular.xyz)*step(0.0, nVP)*step(0.0, lVP);
back_ss = (pow(-lVP, _gl4es_BackMaterial_shininess)*_gl4es_BackLightProduct_0.specular.xyz)*step(0.0, -nVP)*step(0.0, -lVP);
Color.rgb += att*(aa+dd+ss);
BackColor.rgb += att*(back_aa+back_dd+back_ss);
// end of light 0
// light 1 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_1.position.xyz);
aa = _gl4es_FrontLightProduct_1.ambient.xyz;
back_aa = _gl4es_BackLightProduct_1.ambient.xyz;
nVP = dot(normal, VP);
dd = nVP * _gl4es_FrontLightProduct_1.diffuse.xyz*step(0.0, nVP);
back_dd = -nVP * _gl4es_BackLightProduct_1.diffuse.xyz * step(0.0, -nVP);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (pow(lVP, _gl4es_FrontMaterial_shininess)*_gl4es_FrontLightProduct_1.specular.xyz)*step(0.0, nVP)*step(0.0, lVP);
back_ss = (pow(-lVP, _gl4es_BackMaterial_shininess)*_gl4es_BackLightProduct_1.specular.xyz)*step(0.0, -nVP)*step(0.0, -lVP);
Color.rgb += att*(aa+dd+ss);
BackColor.rgb += att*(back_aa+back_dd+back_ss);
// end of light 1
Color.a = _gl4es_FrontMaterial_alpha;
Color.rgb = clamp(Color.rgb, 0., 1.);
BackColor.rgb = clamp(BackColor.rgb, 0., 1.);
BackColor.a = _gl4es_BackMaterial_alpha;
// texturing
}
````

Fragment:
```
#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000000, texture=00000000, texformat=00000000 point=0
varying vec4 Color;
varying vec4 BackColor;
varying bool _gl4es_FrontFacing;
void main() {
vec4 fColor = (_gl4es_FrontFacing)?Color:BackColor;
gl_FragColor = fColor;
}
```

**Comment by kas1e** [comment]
Something wrong with both shaders from last post : they even didn't compiles via glslangvalidator (kronos's tool for checking shaders).

Fragment one ERROR: 0:9: 'in' : cannot be bool

Vertex: 'out' cannot be bool
'gl_FronFacing' : undeclared identifier
'assing' cannot convert from 'temp float' to 'smooth out bool'

Its not on amigaos even, just pure glslangvalidator

**Comment by ptitSeb** [comment]
A varying cannot be a bool :(

I'm sorry, but I don't see any practical way to code a workaround in that case (because of the other bug that prevent me to do `if` in the vertex shader).

**Comment by kas1e** [comment]
Damn :) If it only was in ogles driver which author very fast on fix bugs .. but its in warp3d , which authors .. _not_ fast :) 

I will make bug report about "Loading from built-in variables is unsupported in the fragment shaders". Or maybe not only in fragment, but in any ..

**Comment by ptitSeb** [comment]
On the other hand, this bug is only blocking for two-sided lighting. I'm not sure many games use that feature, so it should block the SDL issue investigation.

**Comment by kas1e** [comment]
You mean "shouldn't" probably ?:) Yes, you are right.
I will create BZ report for that issue, and a bit persists on the needs for fast fix of "if freeze" issues at least. So we can deal with that too.

If you doesn't mind, let it be opened, so i will not forget about needs to ask developers about fixes again and again :)

**Comment by ptitSeb** [comment]
Yeah, "shouldn't". I edit my post :( (so many typo I can do...).

Sure, keep the issue open, we'll close it when you have the bugfix.

**Comment by ptitSeb** [comment]
In fact, it was some left-over some GLES1 conversion in the source of blobloats I used. I removed it and the color are fine now.

I guess the BGRA->RGBA conversion is not bigendian compatible (make sense of course).
I assume you have `__BIG_ENDIAN__` automatically define on your side?

**Comment by ptitSeb** [comment]
I tried something for the BGRA->RGBA conversion on BigEndian. Is it better on your side now?

**Comment by kas1e** [comment]
With that change, things definately changes, but they still not right, new screenshot:
http://kas1e.mikendezign.com/aos4/gl4es/games/bloboads/endian1.jpg
But its definately the right place to change, and we in the right direction there for sure.

Also, i recieve new warp3d driver, where author together with outher fixes (to be tested) added gl_FrontFacing and gl_FragCoord. So, i check our fargment shader which we have from that Morph3d example:

#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000000, texture=00000000, texformat=00000000 point=0
varying vec4 Color;
varying vec4 BackColor;
void main() {
vec4 fColor = (gl_FrontFacing)?Color:BackColor;
gl_FragColor = fColor;
}

And while that shader compiles and checks fine by our internal shader compiler checker (what mean it is indeed added), it still fails inside of GL4ES like that:

glCompileShader(257)
glGetShaderiv(257, 0x8B81, 0x6b0564b4)
glGetShaderInfoLog(257, 1000, 0x0, 0x6b0564b8)
LIBGL: FPE Fragment shader compile failed: ERROR: Code generation failed.
Error log:
OpSelect instruction isn't implemented yet.
INTERNAL ERROR: Store instruction's source parameter is incomplete:
	            OpStore: : tmp21 >> gl_FragColor

Is there still anything missed we need to add, or it's GL4ES side ? (through as it work internally by our checker tools, that probably gl4es ?)

**Comment by ptitSeb** [comment]
Mmmm, for the BGRA->RGBA, I'll try to fix it (all those big/little endian byte moving is really a nightmare).

For the Shader, it's not gl4es, the error message came from the shader compiler. So I'm not sure why the validator works but not the actual compiler. But there is nothing to be done in gl4es here :( It seems (but I'm really not sure at all) that the ternary operator on the builtin doesn't really work, and `fColor` is not filled correctly? You should send the error log the Warp3D developper, he would know better then me what's going on here.

**Comment by kas1e** [comment]
Hm.. Strange that validator works, and not in code.. It can be that language wise code works, but not fill right arguments ?

**Comment by ptitSeb** [comment]
Compilation of a shader is a simple thing. You send the sources to the driver and you compile it.  Arguments coherency between fragment and vertex are checked only at the Link of the program, not at the compilation of the Shader. Really, all is done in the Driver on this stage.

Also, I have just pushed some more changes on endiannes for textures.

**Comment by kas1e** [comment]
Sended all info to warp3d developer

Checked endian fixes : yeah ! 
Take a look :):
http://kas1e.mikendezign.com/aos4/gl4es/games/bloboads/endian2.jpg
http://kas1e.mikendezign.com/aos4/gl4es/games/bloboads/endian3.jpg

Btw, i also recieve in new driver fix for ternary stuff, so will check this one as well.

Now will try that tetrisclone, which black screen for me on gl4es, but works on legacy. You may want to port it to pandora as well: http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/lettersfall-src_beta2.99a.zip

So we can see if it gl4es issue, or, as usuall, on our side :)

**Comment by ptitSeb** [comment]
Good for endianess texture :)

Too bad there is no fps counter on bloboats, because I think gl4es would show good performances against minigl.

For letterfall, I'll do a pandora build and see how the rendering is done.

**Comment by ptitSeb** [comment]
letterfall works fine on my side.
No issue at all.

And the fps in game is 59, so I guess it really is 60fps with vsync...

**Comment by kas1e** [comment]
I need rebuild the letterfall against new gl4es with endian fixes, maybe it will help somehow. Also will try to debug it futher to see where issue arise.

**Comment by ptitSeb** [comment]
Also, for bloboats, press `F2` ingme to show fps counter.

But, not sure you will see much differences. On the Pandora, I have 60fps (using GLES1.1 backend, and limited by vsync here) while also compiling an emulator at the same time...

**Comment by ptitSeb** [comment]
Another game you can try is Critical Mass: http://criticalmass.sourceforge.net/critter.php
OpenGL renderer is pretty simple too.

Also naev.org was working fine. My last test was with version 0.6.1, I have to try with the latest 0.7.0.

**Comment by kas1e** [comment]
Tryed blobots fps counter : minigl : 123fps, gl4es with debug output enabled, but redirected to the null: 94 fps. I need to rebuild it without debug, to see if it make any differences.

**Comment by ptitSeb** [comment]
Yes, debug, with all those printf, will slow things down a bit, even when redirected to null.

I'm revisiting my ported game library, looking at OpenGL one.

Another one you can try (in case you don't have enough title to port :p ) is Cylindrix: https://github.com/hyperlogic/cylindrix

**Comment by kas1e** [comment]
Without debug it give 113fp. So, only on 10fps less than legacy minigl. Through, dunno, if it should be like this anyway ?

Our ogles2 authors says that:

It's no surprise that MGL is faster if the TCL workload is low.  As soon as MiniGL spends more time in its software TCL than GL4ES need to eventually update the internal VBOs / DBOs / whatever-bookkeeping, MiniGL will become slower.  However, at some other point, if the workload is high enough so that the time spent by the GPU outweights the time spent in the respective library significantly, I expect GL4ES to give more or less similar results.

**Comment by ptitSeb** [comment]
Yeah, bloboats is not GPU intensive. You need more complex game ;)

**Comment by kas1e** [comment]
Firstly want to deal with letterfall, as it cleary show that we have bug somewhere (or in my hacked SDL1 , or in our OGLES2  driver). On minigl version with SDL1 it works fine. And as you say it works fine for you too. So probably again ogles2 driver.. But then how to debug it ..

**Comment by kas1e** [comment]
SDL_GL_SwapBuffer() definatelly calls (i put prinfs in, and i can see them).
Is log showing anything bad maybe ? http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/letters_fall_debug.txt

**Comment by kas1e** [comment]
I added some simple quad drawing,  in the visuals.cpp, right after video mode creates:

	glClearColor(0.4f, 0.3f, 0.3f, 255.0f);
	glClear(GL_COLOR_BUFFER_BIT);

	glBegin(GL_QUADS);
		glColor3f(0.7f, 0.8f, 0.3f);
		glVertex2f(-0.5f, 0.5f);
		glVertex2f(0.5f, 0.5f);
		glVertex2f(0.5f, -0.5f);
		glVertex2f(-0.5f, -0.5f);
	glEnd();
	
	glFlush(); 

	SDL_GL_SwapBuffers();

And it draws, then all loads, and once main loop starts, everything going black and non-visibly, invistigate..

**Comment by kas1e** [comment]
Hm, very intersting.. If i put the same test block, to the main.cpp, right before words  printf("Main loop...\n");, also add SDL_Delay(300); so to see what happen, then, i can see how screen clears by my color, but ! no quad draws ! Like glBegin/glEnd didn't take in account anymore ! And then after pause all clear by black, and nothing visibly.. Hm !

**Comment by ptitSeb** [comment]
There is something odd in the log.
If you look for " of 63": that's the fps counter, and this should happens at the SwapBuffers... The fps counter is always "0", that's very low, not sure it's really normal.

All drawing commands seems to be here...

**Comment by ptitSeb** [comment]
You can also try to press "d" to activate some debug info.

**Comment by kas1e** [comment]
Its normal, i run for now Legacy compiled version, it  also shows the same 0 63 at begining. 
And once everything should draws at once, it just draws, while in gl4es version it also draws, just i see nothing but only black screen (but i can click by mouse on menu, etc, i.e. all logic, etc , works, just don't see it)

**Comment by ptitSeb** [comment]
I think the "main" drawing command is in `screens.cpp`.
Line 98 there is a "global" `glColor4f(1.0f, 1.0f, 1.0f, ScreenFadeTransparency);` that (I think) control the color of the screen, for the FadeIn / FadeOut effect.
You can try to play a bit with this. Change `ScreenFadeTransparency` to `1.0f` or move that line inside the `glBegin()`/`glEnd()`.

**Comment by kas1e** [comment]
Tried like this:

			glBegin(GL_QUADS);
                        glColor4f(1.0f, 1.0f, 1.0f, 0.5f);//ScreenFadeTransparency);
			glEnd();

Didn't change anything .. straaange

**Comment by ptitSeb** [comment]
No, not like that.
Here you draw nothing, it's normal nothing change

Change Lines 103-115 to
```
            glBegin(GL_QUADS);
                glColor4f(1.0f, 1.0f, 1.0f, ScreenFadeTransparency);
                glTexCoord2i(0, 0);
                glVertex3i(0-32, 0-32, 0);

                glTexCoord2i(1, 0);
                glVertex3i(0+32, 0-32, 0);

                glTexCoord2i(1, 1);
                glVertex3i(0+32, 0+32, 0);

                glTexCoord2i(0, 1);
                glVertex3i(0-32, 0+32, 0);
            glEnd();
````

**Comment by kas1e** [comment]
I even comment out whole GL fucntions (starting from glPushMatrix(); and ends with glPopMatrix();, and put there test-block which clear background and should drawn quad : backround clears, but quad didn't draws.

**Comment by kas1e** [comment]
Your way also didn't change it .. can it be some alpha channel problems for example, which need to be translated to the big endian or something of that sort ?

**Comment by ptitSeb** [comment]
When you draw your test quad, you also need to disable Texturing (and re-enabling after): put `glDisable(GL_TEXTURE_2D);` before the `glBegin(...)` and a `glEnable(GL_TEXTURE_2D);` after the `glEnd()`.

**Comment by ptitSeb** [comment]
Also, I'm not sure of the state of the ClientState arrays, so to be safe put also a 
```
    glDisableClientState( GL_VERTEX_ARRAY );
    glDisableClientState( GL_TEXTURE_COORD_ARRAY );
```
before the `glBegin(...)`

**Comment by ptitSeb** [comment]
But I really don't see what could be wrong. I don't think it can be a bigendian issue here. The texture are of the same type as with Bloboats, with mixed RGBA and BGRA / UBYTE.
Only thing I see are:
The of glBlend (not sure it was use in Bloboats), but that should work, I don't any reason it wouldn't
The game call pretty often `glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, ...)` and `glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, ...)` before blitting some texture, to set both MIN/MAG filter to GL_LINEAR or GL_NEAREST. While it's useless because it as already be done, maybe some driver doesn't like that (like in `Visuals::DrawLoadedTextOntoScreenBuffer` from visuals.cpp`).

**Comment by kas1e** [comment]
Tried to put :

glDisableClientState( GL_VERTEX_ARRAY );
glDisableClientState( GL_TEXTURE_COORD_ARRAY );

Before glBegin() in the screens.cpp / ApplyScreenFadeTransition: didn't help.

I will try for now  to just remove step by step big chunks of game's code, till i will not shorten it to the maximum possible. Then, we will find the roots for sure.

**Comment by ptitSeb** [comment]
Try to disable Blending also, for testing purpose
add
`glDisable(GL_BLEND);`
for example.

**Comment by ptitSeb** [comment]
When this bug is fixed, another game that I have just test and is working fine with gl4es and GLES2 backend is Yamagi Quake 2: https://github.com/yquake2/yquake2

**Comment by kas1e** [comment]
gldisable(GL_BLEND); didn't change behaviour ..
I almost reduce everything to minimum, probably soon will limit it to some small example

**Comment by kas1e** [comment]
Check this out: http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/main.cpp

That is what i do with main now. As you see, it is only do init game (that call SDL_INIT + some gl calls), then load all stuff , and then only small block which do:

		glClearColor(70, 100, 40, 255);
		glClear(GL_COLOR_BUFFER_BIT);

		
        visuals->DrawTextOntoScreenBuffer("Bringing back old memories from the 16bit era!", visuals->FontDefault[1]
                                          , 0, 307, JustifyCenter, 1, 255, 1, 1, 1, 1);
		glFlush();
		SDL_GL_SwapBuffers();

		SDL_Delay(5000);

And then exit. I.e. draw just some words.

On GL4ES vesion even that didn't works. I.e. i can see background color, but it didn't draws. So.. that lead us to the GL calls in the DrawTextOntoScreenBuffer(), which in the visuals.cpp, and have good bunch of gl calls related to texture loading, etc.  (see after SDL_BlitSurface(text, 0, textAndOutline, &rect); .  Probably (i am not sure, but very possible) : something there. There and glblend, and glGenTextures, and glTexImage2D (that one, looks pretty possible that can make something wrong), glBindTexture , glEnableClientState , glVertexPointer, glDrawArrays and co. So i assume it is something there.

**Comment by ptitSeb** [comment]
I nothing special in that `DrawTextOntoScreenBuffer` method. All is fairly regular operation.
Let's try to pinpoint the issue.

First, lets remove the texture, and just show some white rectangle:

Starting Line 754 of `visuals.cpp`:
```
    glDisable(GL_TEXTURE_2D);//glEnable(GL_TEXTURE_2D);
    //glBindTexture(GL_TEXTURE_2D, textTexture);
    glColor3f(1.0f, 1.0f, 1.0f);

    glTranslatef(posX, posY, 0.0f);

    GLfloat zeroWidth = 0, oneWidth = 0;
    GLfloat zeroHeight = 0, oneHeight = 0;

    zeroWidth = 0; oneWidth = w;
    zeroHeight = 0; oneHeight = h;

    int numvertices = 4;
    glEnableClientState( GL_VERTEX_ARRAY );
    glDisableClientState( GL_TEXTURE_COORD_ARRAY );//glEnableClientState( GL_TEXTURE_COORD_ARRAY );

    GLfloat *vertices = new float[3*numvertices];
    GLfloat *textcoord = new float[2*numvertices];

    textcoord[0] = 0;
    textcoord[1] = 0;
    textcoord[2] = 1;
    textcoord[3] = 0;
    textcoord[4] = 1;
    textcoord[5] = 1;
    textcoord[6] = 0;
    textcoord[7] = 1;

    vertices[0] = zeroWidth;
    vertices[1] = zeroHeight;
    vertices[2] = 0;
    vertices[3] = oneWidth;
    vertices[4] = zeroHeight;
    vertices[5] = 0;
    vertices[6] = oneWidth;
    vertices[7] = oneHeight;
    vertices[8] = 0;
    vertices[9] = zeroWidth;
    vertices[10] = oneHeight;
    vertices[11] = 0;

    glVertexPointer( 3, GL_FLOAT, 0, vertices );
    //glTexCoordPointer( 2, GL_FLOAT, 0, textcoord );

    glDrawArrays( GL_TRIANGLE_FAN, 0, numvertices );
    delete textcoord;
    delete vertices;

    glPopMatrix();
```

**Comment by ptitSeb** [comment]
Also, just to be on the safe side, on your main.cpp, add
`glDisable(GL_DEPTH_TEST);`  before the `printf("Main loop...");`

**Comment by kas1e** [comment]
With that changed code, with or without glDisable(GL_DEPTH_TEST) in the main : on legacy minigl in both cases i have white quad in down of window, in gl4es : always black screen, no quad ! Very strange !

**Comment by kas1e** [comment]
We can try to remove from main call to the DrawTextOntoScreenBuffer at all, and try to draw quad at least there ?

**Comment by kas1e** [comment]
Well, that start to be more strange than ever !

That how looks like my current main.cpp :
http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/main2.cpp

As you can see, i didn't call DrawTextOntoScreenBuffer, and instead, trying to draw QUAD. With gldisable/enabled of texture2d . And , it draw nothing !

**Comment by ptitSeb** [comment]
does this simplified main show a white quad with minigl?

**Comment by kas1e** [comment]
With http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/main2.cpp  : tested. nope, also didn't show. So probably glDisable/glEnabled of 2d texturing not enough here (probably some blend stuff involved too, or something). 

So we can back on visuals.cpp then, where on minigl we have white quad, but on gl4es not and start from there again..

**Comment by ptitSeb** [comment]
No I think it don't show because the matrix are not the same.

`Visuals::SetupGameWindow` do a `glOrtho(0, 640, 480, 0, -1, 1);` on the Projection matrix..

**Comment by kas1e** [comment]
Ok, that works in gl4es:

	glDisable(GL_TEXTURE_2D);	
	
	glClearColor(0.4f, 0.3f, 0.3f, 255.0f);
	glClear(GL_COLOR_BUFFER_BIT);

	glBegin(GL_QUADS);
		glColor3f(0.7f, 0.8f, 0.3f);
		glVertex2f(50, 300);
		glVertex2f(300, 300);
		glVertex2f(300, 50);
		glVertex2f(50, 50);
	glEnd();
	glEnable(GL_TEXTURE_2D);


I also tryied to just add it to not main (it works there), but to the visuals.cpp where we do changes to see white rectangle (and which we don't in gl4es version).  There is also works if i add it right after SDL_BlitSurface(text, 0, textAndOutline, &rect); (lin 737).

And if i add that block right after glPopMatrix(); at end of fucntions, it works too.

But when we do changes about which you told, can't see white quad in gl4es version..

**Comment by ptitSeb** [comment]
So, the `glDrawArrays` doesn't work?

What if change
```
    glVertexPointer( 3, GL_FLOAT, 0, vertices );
    //glTexCoordPointer( 2, GL_FLOAT, 0, textcoord );

    glDrawArrays( GL_TRIANGLE_FAN, 0, numvertices );
```

with
```
    //glVertexPointer( 3, GL_FLOAT, 0, vertices );
    //glTexCoordPointer( 2, GL_FLOAT, 0, textcoord );

    //glDrawArrays( GL_TRIANGLE_FAN, 0, numvertices );
    glDisableClientState( GL_VERTEX_ARRAY );
    glBegin(GL_TRIANGLE_FAN);
     glVertex3f(zeroWidth, zeroHeight, 0.f);
     glVertex3f(oneWidth, zeroHeight, 0.f);
     glVertex3f(oneWidth, oneHeight, 0.f);
     glVertex3f(zeroWidth, oneHeight, 0.f);
    glEnd();
```

**Comment by kas1e** [comment]
That one works, yes !

**Comment by ptitSeb** [comment]
And add the texture now.

```
    //glVertexPointer( 3, GL_FLOAT, 0, vertices );
    //glTexCoordPointer( 2, GL_FLOAT, 0, textcoord );

    //glDrawArrays( GL_TRIANGLE_FAN, 0, numvertices );
    glDisableClientState( GL_VERTEX_ARRAY );
    glEnable(GL_TEXTURE_2D);
    glBegin(GL_TRIANGLE_FAN);
     glTexCoord2f(0, 0);
     glVertex3f(zeroWidth, zeroHeight, 0.f);
     glTexCoord2f(1, 0);
     glVertex3f(oneWidth, zeroHeight, 0.f);
     glTexCoord2f(1, 1);
     glVertex3f(oneWidth, oneHeight, 0.f);
     glTexCoord2f(0, 1);
     glVertex3f(zeroWidth, oneHeight, 0.f);
    glEnd();
```

**Comment by ptitSeb** [comment]
Ooops, closed by mistake.

Reopenned....

**Comment by kas1e** [comment]
Yeah ! Replaced in original sources of whole games all parts where it was, and all can be visibly now ! 
Through have little glitch when come to the "options" menu. Its a bit messed there and there , can you check pl if you have that ? 

So... then, what it did mean ? It mean that glDrawArray didn't works ? In any condition ? 
Is it possible to create very simple test case with glDrawArray to see if it just broken in whole ?

**Comment by ptitSeb** [comment]
Minimum test, take you main2.cpp and change
```
	glBegin(GL_QUADS);
		glColor3f(0.7f, 0.8f, 0.3f);
		glVertex2f(-0.5f, 0.5f);
		glVertex2f(0.5f, 0.5f);
		glVertex2f(0.5f, -0.5f);
		glVertex2f(-0.5f, -0.5f);
	glEnd();
```
with
```
    int numvertices = 4;
    glEnableClientState( GL_VERTEX_ARRAY );

    GLfloat *vertices = new float[3*numvertices];

    vertices[0] = -0.5f;
    vertices[1] = 0.5f;
    vertices[2] = 0;
    vertices[3] = 0.5f;
    vertices[4] = 0.5f;
    vertices[5] = 0;
    vertices[6] = 0.5f;
    vertices[7] = -0.5f;
    vertices[8] = 0;
    vertices[9] = -0.5f;
    vertices[10] = -0.5f;
    vertices[11] = 0;

    glVertexPointer( 3, GL_FLOAT, 0, vertices );

    glDrawArrays( GL_QUAD, 0, numvertices );

    delete vertices;
```

**Comment by kas1e** [comment]
Well, before i tring your test, i even do found one of my simple test cases, where i can use glDrawArray : and, it serusly broken. Take a look:

http://kas1e.mikendezign.com/aos4/gl4es/glDrawArrays/texsub_gl4es.c

Over minigl:
http://kas1e.mikendezign.com/aos4/gl4es/glDrawArrays/gldrawarray_mgl.jpg

Over gl4es:
http://kas1e.mikendezign.com/aos4/gl4es/glDrawArrays/gldrawarray_gl4es.jpg

At least show something, but seriusly broken !

**Comment by kas1e** [comment]
With your final test : and gl4es and minigl versions show nothing ..

**Comment by kas1e** [comment]
But that probably not so matter for now, as my test case show pretty well that something is broken

**Comment by ptitSeb** [comment]
I have a little theory: here we feed Vertex array with 3 float per vertex (x, y, z), but the shader use vec4 (x, y, z, w). The GLES2 driver should make w=1.0f for all vertex. But maybe not with GLES2 / Warp3D.

So let's make a 4 floats with:
```
	float array[] = 
	{
		0.0, 0.0, -2.0, -1.0, 0.0, 1.0f,
		0.0, 1.0, -2.0, 1.0, 0.0, 1.0f,
		1.0, 1.0, 0.0, 1.0, 0.0, 1.0f,
		1.0, 0.0, 0.0, -1.0, 0.0, 1.0f,
 
		0.0, 0.0, 1.0, -1.0, 0.0, 1.0f,
		0.0, 1.0, 1.0, 1.0, 0.0, 1.0f,
		1.0, 1.0, 2.41421, 1.0, -1.41421, 1.0f,
		1.0, 0.0, 2.41421, -1.0, -1.41421, 1.0f
	};
	
	GLubyte index[] =
	{
		4, 5, 6, 7, 0, 1, 2, 3
	};
	
			
   glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
   glEnable(GL_TEXTURE_2D);

   glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_REPLACE);
   glBindTexture(GL_TEXTURE_2D, texName);

   glInterleavedArrays(GL_T2F_V4F, 0, array);
```

(I didn't tested my minitest, lets drop it)

**Comment by kas1e** [comment]
You mean last one to check inside of the game's code, or just make little test case with that code ? As GL_T2F_V4F and texName not declared..

**Comment by ptitSeb** [comment]
Change that code in http://kas1e.mikendezign.com/aos4/gl4es/glDrawArrays/texsub_gl4es.c

**Comment by kas1e** [comment]
Yes, now it looks correct with that code.

So , your theory right then ?

**Comment by ptitSeb** [comment]
Yes it seems.
You need to discuss this problem with OGLES2 driver guy first I guess, and then maybe the Warp3D driver guy, depending on what OGLES2 driver do with vertexattrib...

**Comment by kas1e** [comment]
Uh, that one was hard !

Btw, is Letters Fall works for you normally ? I mean, for me, it have some issues (after i change everywhere code of glDrawArrays on your add-texture one).

For example "menu options" looks like this:
http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/menu.jpg

And in game also something seems broken:
http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/ingame.jpg

**Comment by ptitSeb** [comment]
I have no such issue. There must be some issue in the conversion glDrawArrays -> glBegin/glEnd.
Maybe wait for the fix if it's not too long.

**Comment by kas1e** [comment]
Yeah, let's wait for fix. just probably it will be not 1-day-fast one,as Daniel (osgles2 author), says that he of course can add it at very begining, but didn't do it , simply because of efficiency: if your shader matches your vertex-setup, then adding a potentially ton  of extra data which is totally useless in such a case, kills a huge  optimization opportunity.

So, we ask author of Warp3D: if he can / would implement it on the GPU, as it may well be that it may result in a performance penalty on the GPU. And if that will be the case, then Daniel will add it to OGLES2, just with ability to disable if it need it for someone who do not want to loose any bit of perfomance. 

So will see it ..

In meantime, i may try to debug Cadog . Its show me the same white screen even on legacy opengl now, but i remember i port it 5 years ago, and fix something.. Need to invistigate a bit

**Comment by ptitSeb** [comment]
It's in the OpenGL specs... Vertex Attribs are always vec4, whatever the type of data you set, or the type of data you read in the vertex shader.

I can also make a temporary copy in gl4es, but, as Daniel said, it would not be very efficient, so it would be better to have the Warp3D driver taking care of that.

Still, I can add a quick workaround, at least for the Vertex coord, for amigaos. That will allow you to continue testing, and will probably fix Cadog as well (as it runs fine on my side with gl4es as-is).

**Comment by ptitSeb** [comment]
@kas1e : I have pushed a workaround. It should allow you to make cadog run without touching the code.

Also, I have also pushed another commit (before) to have the Hardware capability check run when a new context in Binded. That means OGLES2 driver capabilities will be queried instead of using minimum spec. So maximum texture res, number of attributes, or some extension will be detected and used (I'm interested in seeing what gl4es sees on the Amiga).

**Comment by kas1e** [comment]
If it not very annoing, we can later just remove it once recieve fix for. 

As for Cadog, i rebuild just all from scratch, etc, and can run it ! Through, have some issues which i didn't have in legacy opengl version. Take a look at screenshots:

http://kas1e.mikendezign.com/aos4/gl4es/games/cadog/1.jpg
http://kas1e.mikendezign.com/aos4/gl4es/games/cadog/2.jpg

You can see on first screenshot, that words "play game" , "settings", "about" and "exit", not missed, they just not filled by yellow color and looks a bit like "transparently".  I mean if you take a close looks at them, you will see that words is here, just not filled right by color.

And on second is that what i have when press on "play game", so probably that one cleary endian one.

And then, once i press it all and so game level1 should starts, it just crash, in the "AmiglDrawArrays" :)

**Comment by kas1e** [comment]
Will check now your changes from previous comment (i just missed it while write my post)

**Comment by ptitSeb** [comment]
Mmmm, yes, the texture issue might be an endianess issue. I need to see what texture format/type/internalformat it tried to use.

**Comment by kas1e** [comment]
Have error when compiling latest commits:

src/glx/hardext.c: In function 'GetHardwareExtensions':
src/glx/hardext.c:132: warning: implicit declaration of function 'gles_glGetString'
src/glx/hardext.c:132: warning: initialization makes pointer from integer without a cast
src/glx/hardext.c:197: warning: implicit declaration of function 'gles_glGetIntegerv'
src/glx/hardext.c:225: warning: implicit declaration of function 'egl_eglQueryString'
src/glx/hardext.c:225: error: 'eglDisplay' undeclared (first use in this function)
src/glx/hardext.c:225: error: (Each undeclared identifier is reported only once
src/glx/hardext.c:225: error: for each function it appears in.)
src/glx/hardext.c:225: error: 'EGL_EXTENSIONS' undeclared (first use in this function)

**Comment by ptitSeb** [comment]
Ah yes, sorry. It should be fixed now.

**Comment by kas1e** [comment]
Yep, do test it now.  Results are: for Cadog no changes. For LettersFall : i revert our changes to originals, and rebuild with new gl4es: all is black :( Probably we can not waste our time on it, until fix done.

Also,  when i start any app which build with new gl4es, i hae LIBGL: hardware tst disabled, nothing activated ... 

I remember we disable it at the begining of amigaos4 tests , maybe it still disabled ?

As for textures for cadog: i build texture.c with debug, and result there:
http://kas1e.mikendezign.com/aos4/gl4es/games/cadog/debug_texture.txt

**Comment by ptitSeb** [comment]
For hardware testing, it will work only if using aglXXXX function from gl4es. If you create context and make it current with OGLES directly, there will still be no test.

For texture of cadog, Ok, I see how to fix it, it will be easy :)

For the backscreen with the new VERTEX4F workaround, I'm surprised, need to add some debug info here!

**Comment by ptitSeb** [comment]
So, for hardware testing, be sure to call `aglMakeCurrent(context);` from gl4es directly and not `OGLES2->aglMakeCurrent(context);` and you should see `LIBGL: Hardware test on current Context...` (the message about deactivated test will still be there, before this one).

The texture issue in cadog should be fixed now.

The vertex4f workaround should be fixed now too.

**Comment by kas1e** [comment]
For hardware test i have then:

4/0.NO NAME:cadog> cadog_gl4es -w -s 640x480
LIBGL: Initialising gl4es
LIBGL: v1.0.5 built on Feb 27 2018 01:38:02
LIBGL: Using GLES 2.0 backend
LIBGL: OGLES2 Library and Interface open successfuly
LIBGL: Hardware test disabled, nothing activated...
LIBGL: Targeting OpenGL 2.0
LIBGL: Current folder is:NO NAME:cadog
LIBGL: FBO are in core, and so used
LIBGL: PointSprite are in core, and so used
LIBGL: high precision float in fragment shader available and used
LIBGL: Max vertex attrib: 16
LIBGL: Max texture size: 16384
LIBGL: Max Varying Vector: 32
LIBGL: Texture Units: 8(8), Max lights: 8, Max planes: 6




> The texture issue in cadog should be fixed now.
Sadly, no changes.

> The vertex4f workaround should be fixed now too.
Sadly, but letterfall still black :(

**Comment by ptitSeb** [comment]
At least the hardware detection works.  Not a lot of stuff (sadly no Cube Maps for example).

For cadog texture, if you can do an update log
Same for letterfall, a log of at least `fpe.c` :(

**Comment by kas1e** [comment]
lettersfall: http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/new_debug1.txt
cadog: http://kas1e.mikendezign.com/aos4/gl4es/games/cadog/debug_new1.txt

**Comment by ptitSeb** [comment]
I think I fixed the texture issue for cadog.

But for letterfall, I don't understand why it wouldn't work. I see the conversion in the log, and I tested the code on my side, things were working fine with the code. I need to think more about that one. I'll sleep on the issue, maybe I'll think of something.
If you can check on the simplest test case of `glDrawArrays` if the fix is working.

**Comment by ptitSeb** [comment]
In fact, I think I know why the vertex4f workaround doesn't work: because I need to apply the workaround not only Vertex, but also on Color, secondarycolor and all texture coordinates vertexattributes...

**Comment by ptitSeb** [comment]
Ok, I think the workaround should work now. Vertex, Color and all 8 textures are "workarounded" now. Off course, this will introduce a slight slowdown, so a proper fix at driver level is much better (plus, it's a quick hack and doesn't cover all cases).

**Comment by kas1e** [comment]
Tested latest commits : 

Good news: Cadog renders textures correctly now , yeah ! :) :
http://kas1e.mikendezign.com/aos4/gl4es/games/cadog/good1.jpg

Then, to bad news: it still crashes in glDrawArrays().

Then, i checked LettersFall: still black screen.

Then, i checked my simple test case, and , from different runs it reacts different, but most of time it looks like this:  http://kas1e.mikendezign.com/aos4/gl4es/glDrawArrays/vertex_wrong1.jpg
But should be like this: http://kas1e.mikendezign.com/aos4/gl4es/glDrawArrays/gldrawarray_mgl.jpg

Imho, don't bother with that Vertex prolem we have. They (authors of ogles2/warp3d) will fix it.

ps. Btw, can you also remove "ternary" fix code from gl4es ? I already recieved a fix for, and all works fine. The less workorounds gl4es have, the better :)

**Comment by ptitSeb** [comment]
Mmmm, so my workaround doesn't work? I had another idea for it anyway, while waiting for the proper fix.

For the ternary workaround, I can remove it, but not sure the workaround is slower then original code. This need to be tested.

**Comment by kas1e** [comment]
As for ternary one: if ternary one can't be faster (i.e. be the same, or slower) : then we do not need it of course. But if it faster, then we can replace main code with it. All in all drivers fixed, and when i do compile things, i always had to remove that amigaos4 ifdef, to avoid building with ternary fix :)

**Comment by ptitSeb** [comment]
Yes the Hardware test are extension for OGLES2 driver (and so maybe from warp3D) that are used by gl4es.
In fact, it seems there is mostly no extensions detected, so that's a bit strange. I was expecting to see a few stuff like `GL_OES_element_index_uint`, `GL_OES_texture_float`, `GL_OES_packed_depth_stencil` or maybe even `GL_EXT_frag_depth`

**Comment by kas1e** [comment]
Maybe those extensions out of plain gles2 specification, so they wasn't implemented in our driver at all.. But i will ask author, maybe he will explain it a bit.

**Comment by ptitSeb** [comment]
It's extension, they are out of main gles2 spec. But I was expecting some of them. It's ok if they are not there. But better if they are present.

**Comment by ptitSeb** [comment]
Latest commit should work now, there is a new wokaround that doesn't copy data...

I also removed the TERNARY workaround for AmigaOS4.

**Comment by kas1e** [comment]
Tried, but vertex shader generation failed:

FPE Vertex shader compiled failed: ERROR: 5:1: '' : syntax error, unexpected IDENIFIER, excting LEFT_BRACE or COMMA or SEMICOLON.

I also have good news from driver's developers, they say it will be implemented in warp3d, so no perfomance penalty will be.

Also always forget to mention: there some missing includes (probably just "math.h"), which make some warning for me about implict declarations. There is list:

src/gl/blit.c: warning: implicit declaration of function 'roundf'
src/gl/fpe.c: warning: implicit declaration of function 'sqrtf'
src/gl/fpe.c: warning: implicit declaration of function 'cosf'
src/gl/init.c:145: warning: implicit declaration of function 'load_libs'
src/gl/init.c:274: warning: implicit declaration of function 'snprintf'
src/gl/line.c:59: warning: implicit declaration of function 'sqrtf'
src/gl/line.c:59: warning: implicit declaration of function 'powf'
src/gl/matrix.c:319: warning: implicit declaration of function 'sqrtf'
src/gl/matrix.c:323: warning: implicit declaration of function 'sinf'
src/gl/matrix.c:324: warning: implicit declaration of function 'cosf'
src/gl/matvec.c:155: warning: incompatible implicit declaration of built-in function 'sqrtf'
src/gl/matvec.c:187: warning: incompatible implicit declaration of built-in function 'sqrtf'
src/gl/pixel.c:1275: warning: implicit declaration of function 'snprintf'
src/gl/program.c: In function 'gl4es_glBindAttribLocation':
src/gl/program.c:60: warning: implicit declaration of function 'strdup'
src/gl/program.c:60: warning: assignment makes pointer from integer without a cast
src/gl/program.c: In function 'gl4es_glLinkProgram':
src/gl/program.c:573: warning: assignment makes pointer from integer without a cast
src/gl/program.c:622: warning: assignment makes pointer from integer without a cast
src/gl/texgen.c:201: warning: implicit declaration of function 'sqrtf'

All others objects compiles warnings free.

**Comment by kas1e** [comment]
Tested that shader which fail now: yes, its even can't be handled by glslangvalidator from kronos, saying:
Error: linking vertex stage: Missing entry point: Each stage reuires one "void main()" entry point.

**Comment by ptitSeb** [comment]
Can you give me the shader the doesn't compile so I can fix the generator.

Also for the warning, I'll see if I can fix them (seems to be mssing some `#include <math.h>` for some of them).

**Comment by kas1e** [comment]
Sure, there is:

[code]
#version 100
precision mediump float;
precision mediump int;
uniform highp mat4 _gl4es_ModelViewProjectionMatrix;
attribute highp vec3 _gl4es_Vertex;
attribute lowp vec1 _gl4es_Color;
attribute highp vec2 _gl4es_MultiTexCoord0;
// FPE_Shader generated
// ** Vertex Shader **
// ligthting=0 (twosided=0, separate=0, color_material=0)
// secondary=0, planes=000000
// texture=00000001, point=0
//    Vertex Attrib size: vertex=3 color=1 secondary=1 textures=00000001
varying vec4 Color;
varying vec2 _gl4es_TexCoord_0;

void main() {
gl_Position = _gl4es_ModelViewProjectionMatrix * vec4(_gl4es_Vertex, 1.0);
Color =  vec4(_gl4es_Color, 1.0);
// texturing
// texture 0 active: 1  
_gl4es_TexCoord_0 = _gl4es_MultiTexCoord0.xy;
} 
[/code]

**Comment by ptitSeb** [comment]
I have pushed a commit that should fix the shader issue now, with the workaround...

About the warning, it seems math.h is included, so where are `roundf` and all other `f` math functions defined on AmigaOS4 if not in `<math.h>`?

**Comment by kas1e** [comment]
As for workaround: tested latest commits, and now it just start crashes on IOGLES2->aglCreateContextTags in my test code.  Probably can be because  of removed load_libs() function or because of changes done in "[GLES2] Refined list of Core function vs Extension in HardExt" commit, dunno.

As for warnings: its all in math.h for us, yep. Maybe it just needs to be included in some other place too, i will check it now on my side.

**Comment by kas1e** [comment]
Ok, found : it is change in the init.c which make it crash now. I.e. line 145 of init.c was load_libs(); , now when that one is removed, we crashes on IOGLES2->aglCreateContextTags in my test code. Probably we need to put that one back.

But anyway, now, after i put "load_libs()" back,  your workoround works on test-case. Then i tried to rebuild Cadog with it : crashe on runing , with words "warnin, gles_GlGetIntegerv is NULL".

**Comment by kas1e** [comment]
Something weird happens in last commits .. not only load_libs are prolems, but something else seems so.

**Comment by kas1e** [comment]
Ok, found : it is change in the init.c which make it crash now. I.e. line 145 of init.c was load_libs(); , now when that one is removed, we crashes on IOGLES2->aglCreateContextTags in my test code. Probably we need to put that one back.

So, do tests:

Simple test case : that one works as should.
LettersFall: works ! through , with the same issues as i describe before (some trashing in options), but in game itself is start to be better.
Cadog: still crasehs in amiglDrawArrays()

But to be honest, you waste your time on that workoround which make no sense. I mean, author of drivers already works on it, soon they will fix it. But you spend your good time on workoround gl4es not need, and which make code of gl4es looks "dirty" and "ifdef" ed a lot :)  Like with case with "ternary" fix : it was good for just fast tests , but is there needs to keep that code in gl4es at all ? (i mean, we only remove ifdef amigaos4 there, but whole "ternary" thing still in code of gl4es).

**Comment by kas1e** [comment]
I think we should wait for a fix imho, to see if the crash in Cadog in glDrawArrays() and trashed font in some parts of LettersFall are related. 

Through, while crash in Cadog may very well be related, i think that issues with LettersFall can be endian ones or so.. I mean take a look at image :

http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/trash_font.jpg

It's just i press on "how to play".

**Comment by kas1e** [comment]
Btw, about your "ternary" workoround keeping in code : we should remove it in whole. It will be not faster than your original way with ternaries. Our ogles2 author explain that this gives the underlying shader-compiler the best hint. The shader-compiler will eventually compile that to a select-operation if appropriate, which the shader-assembler will convert into what best fits the concrete GPU.

In other words, removing it all together will be good idea from code. As well as remove whole vertexattrib4f workoround, and waiting for them to send me fixed drivers :)

**Comment by ptitSeb** [comment]
I don't think the crash in cadog is related to the workaround. It's probably something else. Anyway, it crashes in either OGLES2 or Warp3D driver at this stage.

I have fixed the issue with `load_libs()`, but a crash is `OGLES2->aglCreateContextTags()`? strange, this part hasn't changed.

I stop try to make the workaround working, and I'll remove the code tonight or tomorrow...

**Comment by kas1e** [comment]
When i put load_libs() back, crash in OGLES2->aglCreatecontexttags() gone. I.e. it just wasn't able to open library (because no load_libs() ), so fail on creating context. As we put it back now, it then works again.

As for crash in Cadog: yes, it crashes in the ogles2.library in end. I.e. full crashlog looks like this:
http://kas1e.mikendezign.com/aos4/gl4es/games/cadog/Crashlog_cadog_gl4es_2018-02-27_17-54-34.txt

I.e. in the stacktrace we can see that it going to the ogles2.library and then BUM.

I will forward that to ogles2 developer.

**Comment by kas1e** [comment]
Btw, i also ask Daniel about extensions, there is already few in the code, just wasn't enabled as no one ask for them :)  "Cube Maps" through not supported by Warp3D, so ogles2 as well can't use them. 
GL_OES_element_index_uint , GL_EXT_frag_depth and GL_OES_texture_float - there will be in the next updated drivers probably.

And, want to say you again big fat thanks for working on gl4es and help us with all of this problems !

**Comment by kas1e** [comment]
Ok, Cadog crash on glDrawArrays() fixed by updated ogles2 driver ! There was some issues with internal emulation-VBO for glDrawArrays using client-mem.

Through, it didn't help with LettersFall.. But that one probably better to wait for vertexattribs fix.

**Comment by kas1e** [comment]
Ah, and also extensions enabled now too, that what GL4ES output now:

LIBGL: Initialising gl4es
LIBGL: v1.0.5 built on Feb 27 2018 15:21:42
LIBGL: Using GLES 2.0 backend
LIBGL: OGLES2 Library and Interface open successfuly
LIBGL: Hardware test disabled, nothing activated...
LIBGL: Targeting OpenGL 2.0
LIBGL: Current folder is:NO NAME:cadog
LIBGL: Hardware test on current Context...
LIBGL: Hardware Limited NPOT detected and used
LIBGL: FBO are in core, and so used
LIBGL: PointSprite are in core, and so used
LIBGL: Extension GL_OES_element_index_uint detected
LIBGL: high precision float in fragment shader available and used
LIBGL: Extension GL_EXT_frag_depth detected and used
LIBGL: Max vertex attrib: 16
LIBGL: Max texture size: 16384
LIBGL: Max Varying Vector: 32
LIBGL: Texture Units: 8(8), Max lights: 8, Max planes: 6

**Comment by ptitSeb** [comment]
Yes, some extensions, that's good!
So Cadog runs now, fine.
Unfortunatly, I don't think it's really a GPU intensive game, so again, perf against miniGL will show nothing significant.

For letterfalls, yes, you need the vertexattrib fix.

**Comment by ptitSeb** [comment]
This bug is a bit long at takes time to load... We will have to close it soon.

Next port? Try yquake2 or ioquake3. But you will probably need the vertexattrib bugfix for those to run fine...

**Comment by kas1e** [comment]
ioquake3 probably now sdl2 only, will check yquake2

**Comment by ptitSeb** [comment]
I have a version stiil on sdl1: https://github.com/ptitSeb/ioq3
Also, freespace, also sdl1: https://github.com/ptitSeb/freespace2

But yeah, yquake2 can do both sdl and sdl2 iirc.

**Comment by ptitSeb** [comment]
Note that for yquake2 you may need to use the envrionnement variable LIBGL_MIPMAP=1 to force using mipmap on all texture, or some wall texture maybe missing (I still don't know why, a bug on gl4es side I guess).

**Comment by kas1e** [comment]
What you think if we will that issue open till original issue not fixed, and instead, i will create "quake3 on amigaos4" issue ?

**Comment by ptitSeb** [comment]
Sure, no problem.

**Comment by hunoppc** [comment]
Hello,
Before starting to post this message I introduce myself.
 
I am HunoPPC active developer in the community AMIGA and I am French living on DIJON.
I'm also working on OpenGLES and I'm porting the EGL library
I did a lot of porting and OpenGL work on Amiga
I look very closely at this development GL4ES and find it beautiful the work done, all my congratulations.
For lack of compatible machine OPENGLES here on Paris I will try this weekend to put my stone in the building while trying to operate RTCW and HURRICAN at first to check that there is no problem with the textures.
I'm starting to do tests with GLU and GL4ES too (I'm working on integrating into my EGL LIB to get bored)
Then I'll post my results here
Thank you Kas1e and thank you PtitSeb
cordially
HunoPPC

**Comment by ptitSeb** [comment]
Nice, more testing :)

Hurrican can run on GLES2 I think. 
RTCW have a GLES 1.1 port, but original GL renderer should run without any changes on GL4ES too.

**Comment by ptitSeb** [comment]
Found your homepage @hunoppc . Nice collection of ports!

I see you know those game quite well already.

**Comment by hunoppc** [comment]
Thank you very much, I also return the compliment for all your work
For the realization of Hurrican and RTCW it will be with my native MiniGL AmigaOS4 code that I will gradually adapt to OpenGL 1 it will take a little work but it should do it.
For Hurrican I have an SDL1 version that should do for testing.
If this can move the project forward I'm starting

**Comment by kas1e** [comment]
@ptiSeb
In meantime tested latest ogles2 build, and it seems we have a bit more extensions again:

LIBGL: Hardware test on current Context...
LIBGL: Hardware Limited NPOT detected and used
LIBGL: FBO are in core, and so used
LIBGL: PointSprite are in core, and so used
LIBGL: CubeMap are in core, and so used
LIBGL: BlendColor is in core, and so used
LIBGL: Blend Substract is in core, and so used
LIBGL: Blend Function and Equation Separation is in core, and so used
LIBGL: Extension GL_OES_element_index_uint detected
LIBGL: high precision float in fragment shader available and used
LIBGL: Extension GL_EXT_frag_depth detected and used
LIBGL: Max vertex attrib: 16
LIBGL: Max texture size: 16384
LIBGL: Max Varying Vector: 32
LIBGL: Texture Units: 8(8), Max lights: 8, Max planes: 6 


Through about CubeMaps: ogles2.lib actually contains everything to support cube-maps because it's core GLES2. However Warp3d doesn't support it yet, which is why glGet(GL_MAX_CUBE_MAP_TEXTURE_SIZE) will return 0 although it should not. Once Nova supports it it will work automagically in ogles2.library.

**Comment by kas1e** [comment]
I recieved today new version of warp3d nova, where OpSelect is impemented now, so that test case works fine over gl4es and we can close that ticked :)

**Comment by ptitSeb** [comment]
So, the use of `gl_FrontFacing` is alowed in the Fragment shader now? good! Other stuff got fixed ? (also, I have updated gl4es and fixed many stuff, don't forget to update before testing).

**Comment by kas1e** [comment]
As i understand release notes for for warp3d, gl_FrontFacing was added to warp3d right after we report it, but it also leads to another issues: generated shaders use OpSelect which our shaders didn't have before, so, Hans add that OpSelect stuff some days ago, and now gl_FrontFacing works fine.

As for other issues, together with that one, was fixed only one issue about quake3, when we looks at the mirror (there was some strange dots all around triangles), and that one fixed. Other issues still there, but i need to test latest gl4es as well.
